### PR TITLE
Add the ability to build sdist or wheel package

### PR DESCRIPTION
**Pull request by rgreinho**
Adds the ability to package the models into a sdist or a wheel package
in addition to building the package.

### CODE REVIEW

**Comment by rgreinho** [comment]
This PR should come after https://github.com/jdkandersson/OpenAlchemy/pull/188, due to the testing issues, but as the logic was written I thought I would push the it anyway, so that you can at least already review this part.

I updated the mypy test (`test_generate_type_return()` in `tests/open_alchemy/models_file/test_integration.py`) to display a report if something goes wrong. It helped me debug some typing issues and it does not impact the test itself. 

I would also request your feedback regarding testing some edge cases in the functions I built (`open_alchemy/build/__init__.py`     lines 303-304, 325-326). The goal is to test that an exception is thrown if the `wheel` package is missing or if the execution of the command fails. How did you do to test that `pyyaml` is present in `open_alchemy/__init__.py:init_yaml()`?

**Comment by rgreinho** [comment]
I got the tests to pass but I had to use 4 cheats:

1. Skip bandit warning: https://github.com/jdkandersson/OpenAlchemy/pull/190/files#diff-a6ba5983eec7e784e69d2f6ed3bc5532R7
2. Skip bandit warning: https://github.com/jdkandersson/OpenAlchemy/pull/190/files#diff-a6ba5983eec7e784e69d2f6ed3bc5532R326
3. Xfail unit test: https://github.com/jdkandersson/OpenAlchemy/pull/190/files#diff-5c81ace2ad82907d95a687b1a4be3c8eR502
4. Xfail unit test: https://github.com/jdkandersson/OpenAlchemy/pull/190/files#diff-5c81ace2ad82907d95a687b1a4be3c8eR507

Maybe you have ideas to help me remove them?

File: open_alchemy/__init__.py [hunk 1]
```diff
@@ -250,6 +250,7 @@ def build_json(
     spec_filename: str,
     package_name: str,
     dist_path: str,
+    format_: str = None,
```
Comment (jdkandersson) [inline open_alchemy/__init__.py:253]: Does this need to be `typing.Optional[str]`?

File: open_alchemy/__init__.py [hunk 2]
```diff
@@ -270,13 +271,16 @@ def build_json(
...
+    return _build_module.execute(
+        spec=spec, name=package_name, path=dist_path, format_=format_
+    )
 
 
 def build_yaml(
     spec_filename: str,
     package_name: str,
     dist_path: str,
+    format_: str = None,
```
Comment (jdkandersson) [inline open_alchemy/__init__.py:283]: Does this need to be typing.Optional[str]?

File: open_alchemy/build/__init__.py [hunk 3]
```diff
@@ -218,14 +220,15 @@ def dump(
         raise exceptions.BuildError(str(exc)) from exc
 
 
-def execute(*, spec: typing.Any, name: str, path: str) -> None:
+def execute(*, spec: typing.Any, name: str, path: str, format_: str = None) -> None:
```
Comment (jdkandersson) [inline open_alchemy/build/__init__.py:223]: Does `format` this need to be typing.Optional[str]? Also, given that it defaults to `None` on the interface, does it need to default to `None` here again?

File: open_alchemy/build/__init__.py [hunk 4]
```diff
@@ -246,3 +249,89 @@ def execute(*, spec: typing.Any, name: str, path: str) -> None:
...
+def build_sdist(path: str) -> None:
+    """
+    Build a .tar.gz source distribution and place it in a "dist" folder.
+
+    The "dist" folder is located at the root of the package.
+
+    Args:
+        path: The package directory.
+    """
+    run("python setup.py sdist", path)
```
Comment (jdkandersson) [inline open_alchemy/build/__init__.py:293]: I haven't encountered this before, does this execute a shell command? If so, how can you be sure that the `python` command does what you think it will do? For example, it could point to Python 2.7 on certain systems. Is there a Python API that can produce these?

File: tests/open_alchemy/test_build.py [hunk 5]
```diff
@@ -491,10 +491,27 @@ def test_calculate_version(spec, spec_str, expected_version):
...
 
+@pytest.mark.parametrize(
+    "package_format",
+    [
+        pytest.param(None, id="build without package"),
+        pytest.param("sdist", id="build with a sdist package"),
+        pytest.param(
+            "wheel",
+            id="build with a wheel package",
+            marks=pytest.mark.xfail(reason="the wheel package is not installed"),
```
Comment (jdkandersson) [inline tests/open_alchemy/test_build.py:502]: Could this particular problem be resolve by making the installation also include the `wheel` package?

The additional request I would have is to add a test for when the wheel package is not installed. There should be some examples for when `PyYAML` is not installed in the code based integration tests.

File: open_alchemy/build/__init__.py [hunk 6]
```diff
@@ -246,3 +249,89 @@ def execute(*, spec: typing.Any, name: str, path: str) -> None:
...
+    Build a distibution package.
+
+    Args:
+        path: The package directory.
+        format: The package format to build ("sdist", "wheel" or "all").
+    """
+    formats = {"sdist", "wheel", "all"}
+    fmt = format_.lower()
+    if fmt not in formats:
+        raise ValueError(
```
Comment (jdkandersson) [inline open_alchemy/build/__init__.py:270]: You can avoid this with `Enum`. Also, all exceptions that are raised should inherit from the `BaseError` in `exceptions.py` and be purpose driven. For example, there is already a `BuildError`, do you think that is appropriate here?

**Comment by rgreinho** [comment]
I believe all the concerns have been addressed and this PR is now ready for review. 💪

File: .gitignore [hunk 7]
```diff
@@ -20,4 +22,3 @@ venv
 venv36
 coverage.xml
 docs/source/_*
-.venv
```
Comment (jdkandersson) [inline .gitignore:23]: Is this intentionally removed?

File: CHANGELOG.md [hunk 8]
```diff
@@ -4,6 +4,8 @@
 
 - Refactor the models file generation to use the artifacts from the schemas.
 - Add `build_json` and `build_yaml` interfaces which can be used to produce a package with the models.
+- Ensure a package built with `build_json` or `build_yaml` is contained within a project directory.
```
Comment (jdkandersson) [inline CHANGELOG.md:7]: My preference would be to state what the change is, e.g. `Changed output of `build_json` and `build_yaml` to be contained within a project directory.` As a reader I wouldn't be sure what ensured meant, is it a check that is run?

File: CHANGELOG.md [hunk 9]
```diff
@@ -4,6 +4,8 @@
...
+- Add the ability to build a `sdist` or `wheel` distributable archive.
```
Comment (jdkandersson) [inline CHANGELOG.md:8]: Could this be shortened to "Add support for building `sdist` or `wheel` distributable archive."

File: docs/source/index.rst [hunk 10]
```diff
@@ -137,6 +137,27 @@ installed. The :samp:`build_yaml` interface accepts the following arguments:
 
 * :samp:`dist_path`: The directory to output the package files.
```
Comment (jdkandersson) [inline docs/source/index.rst:139]: Is there intentionally a blank line here? The blank lines above were because there is a numbered list between

File: open_alchemy/__init__.py [hunk 11]
```diff
@@ -8,6 +8,7 @@
 from sqlalchemy.ext import declarative
 
 from open_alchemy import types as oa_types
+from open_alchemy.build import PackageFormat
```
Comment (jdkandersson) [inline open_alchemy/__init__.py:11]: The pattern I have been following is to import at least the module. This makes mocking possible if required for testing. It also helps with namespaces. E.g. change this line to `from open_alchemy import build` and then refer to it using `build. PackageFormat`.

File: open_alchemy/__init__.py [hunk 12]
```diff
@@ -8,6 +8,7 @@
...
 
 from . import build as _build_module
```
Comment (jdkandersson) [inline open_alchemy/__init__.py:13]: Is this a duplicate import? I like the leading `_`, the pattern elsewhere has been just `_build`. Do you think that should be changed?

File: open_alchemy/__init__.py [hunk 13]
```diff
@@ -250,17 +251,26 @@ def build_json(
...
+
+    .. code-block: python
+
+        format_ = PackageFormat.SDIST|PackageFormat.WHEEL
+
     Args:
         spec_filename: filename of an OpenAPI spec in JSON format
         package_name: The name of the package.
         dist_path: The directory to output the package to.
+        format_: The format(s) of the archive to build if any
```
Comment (jdkandersson) [inline open_alchemy/__init__.py:273]: Please note that it is an optional argument

File: open_alchemy/__init__.py [hunk 14]
```diff
@@ -270,24 +280,35 @@ def build_json(
...
+
+    .. code-block:: python
+
+        format_=PackageFormat.SDIST|PackageFormat.WHEEL
+
     Args:
         spec_filename: filename of an OpenAPI spec in YAML format
         package_name: The name of the package.
         dist_path: The directory to output the package to.
+        format_: The format(s) of the archive to build if any
```
Comment (jdkandersson) [inline open_alchemy/__init__.py:311]: Same as above, please note that the parameter is optional.

File: open_alchemy/build/__init__.py [hunk 15]
```diff
@@ -1,8 +1,17 @@
 """Build a package with the OpenAlchemy models."""
-
+from enum import auto
```
Comment (jdkandersson) [inline open_alchemy/build/__init__.py:2]: The pattern elsewhere is to import `enum` so that it can be used as a namespace and to reduce the number of imports. Do you think that makes sense here?

File: open_alchemy/build/__init__.py [hunk 16]
```diff
@@ -23,6 +32,29 @@
...
+    WHEEL = auto()
+
+
+def validate_dist_format(format_) -> None:
+    """
+    Ensure a valid package format is provided.
+
+    Raises a ``BuildError` exception if the format is invalid.
+
+    :param PackageFormat format_: the package format to validate
```
Comment (jdkandersson) [inline open_alchemy/build/__init__.py:49]: There is a pattern for these comments, please follow it. Take a look at the docstyle for `run` below for example :smile:

File: open_alchemy/build/__init__.py [hunk 17]
```diff
@@ -23,6 +32,29 @@
...
+    """
+    if format_ is not PackageFormat.NONE and PackageFormat.NONE in format_:
+        raise exceptions.BuildError(
+            "combining `PackageFormat.NONE` with another package type is "
+            "confusing, please be more specific"
```
Comment (jdkandersson) [inline open_alchemy/build/__init__.py:54]: Might be nitpicking, it would be good to explain why. For example, that `PackageFormat.NONE` indicates that no additional package should be produced but another format does. It would also be good to show the format that was passed in the error to make it obvious.

File: open_alchemy/build/__init__.py [hunk 18]
```diff
@@ -218,16 +250,105 @@ def dump(
...
+
+    .. code-block:: python
+
+        format_=PackageFormat.SDIST|PackageFormat.WHEEL
+
+    Args:
+        path: The package directory.
+        format: The package format to build.
+    """
+    pkg_dir = pathlib.Path(path) / name
```
Comment (jdkandersson) [inline open_alchemy/build/__init__.py:268]: My preference would be to push the calculation of the output path to as late as possible and make `path` and `name` an argument to `build_sdist` and `build_wheel`. If you like, you could add a small utility function that combines them so you don't have to repeat code.

File: open_alchemy/build/__init__.py [hunk 19]
```diff
@@ -218,16 +250,105 @@ def dump(
...
+            cwd=cwd,
+            check=True,
+            shell=False,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+    except subprocess.CalledProcessError as exc:
+        raise exceptions.BuildError(str(exc)) from exc
+    else:
+        return output.stdout.decode("utf-8"), output.stderr.decode("utf-8")
```
Comment (jdkandersson) [inline open_alchemy/build/__init__.py:331]: I would like to understand, what does `else` do here? Would it be simpler if the return statement was in the `try` block or completely out of the exception handling?

File: open_alchemy/build/__init__.py [hunk 20]
```diff
@@ -218,16 +250,105 @@ def dump(
...
+
+
+def execute(
+    *,
+    spec: typing.Any,
+    name: str,
+    path: str,
+    format_: PackageFormat = PackageFormat.NONE,
```
Comment (jdkandersson) [inline open_alchemy/build/__init__.py:339]: Did you make a decision on wanting to keep the default value in place?

File: open_alchemy/build/__init__.py [hunk 21]
```diff
@@ -218,16 +250,105 @@ def dump(
...
     Execute the build for a spec.
 
     Args:
         spec: The spec to execute the build on.
         name: The name of the package.
         path: The build output path.
+        format_: The format of the distribution package to build.
 
     """
+    validate_dist_format(format_)
```
Comment (jdkandersson) [inline open_alchemy/build/__init__.py:351]: 👍

File: setup.py [hunk 22]
```diff
@@ -48,7 +49,6 @@
             "black",
             "pre-commit",
             "isort<5.0.0",
-            "PyYAML",
```
Comment (jdkandersson) [inline setup.py:51]: 👍

File: tests/open_alchemy/test_build.py [hunk 23]
```diff
@@ -3,6 +3,7 @@
 
 from open_alchemy import build
 from open_alchemy import exceptions
+from open_alchemy.build import PackageFormat
```
Comment (jdkandersson) [inline tests/open_alchemy/test_build.py:6]: Is this import needed if `build` is already imported?

File: tests/open_alchemy/test_build.py [hunk 24]
```diff
@@ -585,3 +603,69 @@ def test_execute(tmp_path):
...
+
+
+@pytest.mark.build
+def test_validate_dist_format_valid():
+    """
+    GIVEN a valid format
+    WHEN validate_dist_format is called
+    THEN nothing happens
+    """
+    for format_ in PackageFormat:
```
Comment (jdkandersson) [inline tests/open_alchemy/test_build.py:658]: My preference would be to have this be parametrised. What about the case where both `wheel` and `sdist` are given?

**Comment by jdkandersson** [comment]
Good work :smile:. One other request, could you please update the `README.md` to include the support for building the sdist and wheel? There is already a line for the package build, I propose we add it to that.

File: .gitignore (hunk 7 above)
Comment (rgreinho) [reply .gitignore:23]: Yes, it was already present line 8.

File: CHANGELOG.md (hunk 8 above)
Comment (rgreinho) [reply CHANGELOG.md:7]: Makes sense. Addressed ✅

File: CHANGELOG.md (hunk 9 above)
Comment (rgreinho) [reply CHANGELOG.md:8]: Addressed  ✅

File: docs/source/index.rst (hunk 10 above)
Comment (rgreinho) [reply docs/source/index.rst:139]: Got it. I removed the blank line. ✅

File: open_alchemy/__init__.py (hunk 13 above)
Comment (rgreinho) [reply open_alchemy/__init__.py:273]: Addressed ✅

File: open_alchemy/__init__.py (hunk 14 above)
Comment (rgreinho) [reply open_alchemy/__init__.py:311]: Addressed ✅

File: open_alchemy/build/__init__.py (hunk 15 above)
Comment (rgreinho) [reply open_alchemy/build/__init__.py:2]: Totally! As I like to follow the OpenStack coding guidelines, it is also mentioned there: https://docs.openstack.org/charm-guide/latest/coding-guidelines.html#import-style.

As a side note, maybe it would be worth creating a CONTRIBUTING.md file and add the "rules" you like contributors to follow? As an example, here is such a file that I like to include in my projects: https://github.com/scrapd/scrapd/blob/master/.github/CONTRIBUTING.rst. If you think that could benefit your project, I could create an issue where we could discuss the content.

File: open_alchemy/build/__init__.py (hunk 16 above)
Comment (rgreinho) [reply open_alchemy/build/__init__.py:49]: Got it...  (sorry I am more used to the other way to write docstrings. Old habits die hard 😅 )

File: open_alchemy/build/__init__.py (hunk 17 above)
Comment (rgreinho) [reply open_alchemy/build/__init__.py:54]: Addressed ✅

File: open_alchemy/build/__init__.py (hunk 18 above)
Comment (rgreinho) [reply open_alchemy/build/__init__.py:268]: Got it. I moved it closer to the execution of the `run()` function. I did not create a function just to join a path, it seemed a bit excessive.

File: open_alchemy/build/__init__.py (hunk 19 above)
Comment (rgreinho) [reply open_alchemy/build/__init__.py:331]: It depends on people. As for me I like to use the `else` as the return line should not be part of the try (it will not raise a `CalledProcessError` exception), and I think it helps "gluing" all the pieces of the `try` together (https://realpython.com/python-exceptions/#the-else-clause).

I took it completely out of the exception handling as you don't seem to be a fan of this syntax 😄

File: open_alchemy/build/__init__.py (hunk 20 above)
Comment (rgreinho) [reply open_alchemy/build/__init__.py:339]: Ah yes, I forgot this one. So since our `build_*()` functions already have a default value for the format, I took it away from the `execute()` function.

File: tests/open_alchemy/test_build.py (hunk 23 above)
Comment (rgreinho) [reply tests/open_alchemy/test_build.py:6]: Nope, I removed it ✅

File: tests/open_alchemy/test_build.py (hunk 24 above)
Comment (rgreinho) [reply tests/open_alchemy/test_build.py:658]: Addressed ✅

File: .gitignore (hunk 7 above)
Comment (rgreinho) [reply .gitignore:23]: Yes, it as already present line 8.

**Comment by rgreinho** [comment]
Alright, I think I addressed everything, and I think I changed the `README.md` as you wanted. 👍

File: open_alchemy/build/__init__.py (hunk 15 above)
Comment (jdkandersson) [reply open_alchemy/build/__init__.py]: Very good idea, I will add that as a ticket as well.

File: open_alchemy/build/__init__.py (hunk 19 above)
Comment (jdkandersson) [reply open_alchemy/build/__init__.py]: It was more of a question, I like your reasoning too

**Comment by jdkandersson** [comment]
I have to apologise, I made a mistake in a comment earlier. For people to be able to pass in a `format` to `build_json` or `build_yaml`, they need access to the `PackageFormat` enum. You had imported into the `__init__.py` file for OpenAlchemy but I think I put in a comment suggesting to remove it. I just realised that it means that people would need to use the private module (for example) `_build_module.PackageFormat.WHEEL` to be able to create the package which shouldn't be the case.

**Review by jdkandersson — APPROVED** [review]

**Comment by rgreinho** [comment]
Ah no problem. I replaced the import back to `from open_alchemy import build`.

**Review by jdkandersson — APPROVED** [review]
